src/Tests/ArrayTests.cs

```diff
@@ -41,8 +38,84 @@ public void Should_throw_for_array()
             var exception = Assert.Throws<Exception>(() => serializer.Deserialize(stream, new List<Type>()));
             Approvals.Verify(exception.Message);
         }
+    }
+    [Test]
+    public void Should_throw_for_multiple_passed_type()
+    {
+        var text = @"
+[{
+    SomeProperty: 'Value1'
+},
+{
+    SomeProperty: 'Value2'
+}]";
+
+        using (var stream = new MemoryStream())
+        {
+            var streamWriter = new StreamWriter(stream);
+            streamWriter.Write(text);
+            streamWriter.Flush();
+            stream.Position = 0;
+
+            var messageMapper = new MessageMapper();
+            var serializer = new JsonMessageSerializer(messageMapper, null, null, null, null);
+            var messageTypes = new List<Type> { typeof(ArrayMessage) };
+            var exception = Assert.Throws<Exception>(() =>
+            {
+                serializer.Deserialize(stream, messageTypes);
+            });
+            Approvals.Verify(exception.Message);
+        }
+    }
+
+    [Test]
+    public void Should_not_throw_for_single_dollar()
+    {
+        var text = $@"
+[{{
+    $type: '{typeName}',
+    SomeProperty: 'Value1'
+}}]";
+
+        using (var stream = new MemoryStream())
+        {
+            var streamWriter = new StreamWriter(stream);
+            streamWriter.Write(text);
+            streamWriter.Flush();
+            stream.Position = 0;
```

Comment:
Put the sreamWriter stuff in a helper function for the test project? It just clutters the test code.